crowbar_framework/app/models/provisioner_service.rb

```diff
@@ -177,5 +177,78 @@ def transition(inst, name, state)
     @logger.debug("Provisioner transition: exiting for #{name} for #{state}")
     [200, { name: name }]
   end
-end
 
+  def load_repository_bag
+    bag = Chef::DataBag.load("repositories") rescue nil
+    if bag.nil?
+      create_repository_bag
+    end
+    bag
+  end
+
+  def synchronize_repositories(platforms)
+    # create DataBag if it doesn't exist yet
+    load_repository_bag
+    platforms.each do |platform, repos|
+      repos.each do |reponame, active|
+        repo_object = Crowbar::Repository.where(platform: platform, repo: reponame).first
+        case active.to_i
+        when 0
+          db = Chef::DataBagItem.load("repositories", repo_object.id) rescue nil
+          # skip repo if it hasn't been active before
+          next if db.nil?
+          destroy_repository_item(repo_object)
+        when 1
+          create_repository_item(repo_object)
+        end
+      end
+    end
+  end
+
+  def create_repository_item(repo_object)
+    saved = false
+    # create DataBag if it doesn't exist yet
+    load_repository_bag
+    # create DataBagItem with the repo metadata used by provisioner recipes
+    if repo_object.available? && repo_object.valid_key_file?
+      @logger.debug("Setting #{repo_object.registry['name']} repository for #{repo_object.platform} as active.")
+      repo_object.to_databag!.save
+      saved = true
+    else
+      @logger.debug("Cannot set #{repo_object.registry['name']} repository for #{repo_object.platform} as active.")
+    end
+    saved
+  end
+
+  def destroy_repository_item(repo_object)
+    destroyed = false
+    # create DataBag if it doesn't exist yet
+    load_repository_bag
+    # otherwise destroy DataBagItem
+    db_item = repo_object.bag_item
+    ret = db_item.destroy("repositories", db_item.id) rescue nil
+    if ret
+      @logger.debug("Deactivating #{repo_object.registry['name']} repository.")
+      destroyed = true
+    else
+      @logger.debug("Cannot deactivate #{repo_object.registry['name']} repository.")
+    end
+    # destroy the DataBag if it is empty
+    destroy_repository_bag if load_repository_bag.empty?
+    destroyed
+  end
```

Comment:
Porting the multiple architecture support on top of this: I'm getting convinced it's better to have just one data bag item with everything (more or less mirroring the structure of the yaml file, but just with the attributes that we need). Right now, I would need to change the id to also contain the architecture, which is really weird.

Also https://github.com/crowbar/crowbar-core/pull/18/files#r42978733 is another reason to put everything just in one databag item.